src/wrapper/objects/auto_array.rs

```diff
@@ -0,0 +1,135 @@
+}
+
+// TypeArray builder
+macro_rules! type_array {
+    ( $jni_type:ty, $jni_get:tt, $jni_release:tt ) => {
+        /// $jni_type array access/release impl
+        impl TypeArray for $jni_type {
+            /// Get Java $jni_type array
+            fn get(env: &JNIEnv, obj: JObject, is_copy: &mut jboolean) -> Result<*mut Self> {
+                let internal = env.get_native_interface();
+                let res = jni_unchecked!(internal, $jni_get, *obj, is_copy);
+                Ok(res)
+            }
+
+            /// Release Java $jni_type array
+            fn release(env: &JNIEnv, obj: JObject, ptr: NonNull<Self>, mode: i32) -> Result<()> {
+                let internal = env.get_native_interface();
+                jni_void_call!(internal, $jni_release, *obj, ptr.as_ptr(), mode as i32);
+                Ok(())
+            }
+        }
+    };
+}
+
+type_array!(jint, GetIntArrayElements, ReleaseIntArrayElements);
+type_array!(jlong, GetLongArrayElements, ReleaseLongArrayElements);
+type_array!(jbyte, GetByteArrayElements, ReleaseByteArrayElements);
+type_array!(
+    jboolean,
+    GetBooleanArrayElements,
+    ReleaseBooleanArrayElements
+);
+type_array!(jchar, GetCharArrayElements, ReleaseCharArrayElements);
+type_array!(jshort, GetShortArrayElements, ReleaseShortArrayElements);
+type_array!(jfloat, GetFloatArrayElements, ReleaseFloatArrayElements);
+type_array!(jdouble, GetDoubleArrayElements, ReleaseDoubleArrayElements);
+
+/// Auto-release wrapper for pointer-based generic arrays.
+///
+/// This wrapper is used to wrap pointers returned by Get<Type>ArrayElements.
+///
+/// These arrays need to be released through a call to Release<Type>ArrayElements.
+/// This wrapper provides automatic array release when it goes out of scope.
+pub struct AutoArray<'a: 'b, 'b, T: TypeArray> {
+    obj: JObject<'a>,
+    ptr: NonNull<T>,
+    mode: ReleaseMode,
+    is_copy: bool,
+    env: &'b JNIEnv<'a>,
+}
+
+impl<'a, 'b, T: TypeArray> AutoArray<'a, 'b, T> {
+    /// Creates a new auto-release wrapper for a pointer-based generic array
+    ///
+    /// Once this wrapper goes out of scope, `Release<Type>ArrayElements` will be
+    /// called on the object. While wrapped, the object can be accessed via
+    /// the `From` impl.
+    pub fn new(env: &'b JNIEnv<'a>, obj: JObject<'a>, mode: ReleaseMode) -> Result<Self> {
+        let mut is_copy: jboolean = 0xff;
+        Ok(AutoArray {
+            obj,
+            ptr: {
+                let ptr = T::get(env, obj, &mut is_copy)?;
+                NonNull::new(ptr).ok_or(Error::NullPtr("Non-null ptr expected"))?
+            },
+            mode,
+            is_copy: is_copy == sys::JNI_TRUE,
+            env,
+        })
+    }
+
+    /// Get a reference to the wrapped pointer
+    pub fn as_ptr(&self) -> *mut T {
+        self.ptr.as_ptr()
+    }
+
+    /// Commits the changes to the array, if it is a copy
+    pub fn commit(&mut self) -> Result<()> {
```

Comment:
Does it have to be `&mut`? No modifications to the array/its copy are performed in a release.

Reply:
These are left-overs from the original version. Good catch.